### PR TITLE
Implement a test for proxy.Connection

### PR DESCRIPTION
**Pull request by richvdh**
I feel like, after all this effort, I've ended up with a test which is baroque
and doesn't really test anything except that channels work and that the
websocket library does what it's supposed to.

What do you think?

### CODE REVIEW

**Comment by richvdh** [comment]
retest this please

File: proxy/conn_test.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/matrix-org/matrix-websockets-proxy/mocks/io"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"testing"
+	"time"
+)
+
+// testSendMessage tests that Conn.SendMessage causes a text message to be
```
Comment (illicitonion) [inline proxy/conn_test.go:17]: Name the test with enough information that you don't need a comment?

File: proxy/conn_test.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+// written to the web socket
+func TestSendMessage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	wg := &sync.WaitGroup{}
```
Comment (illicitonion) [inline proxy/conn_test.go:21]: It's a little strange to explicitly make a nil value; I would expect this to just be:

var wg sync.WaitGroup

File: proxy/conn_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+	// make the websocket
+	recvPipeRdr, _ := io.Pipe()
+	writer, ws, err := makeWsConn(recvPipeRdr, ctrl)
+	if err != nil {
+		t.Errorf("Expected no error, got '%v'", err)
```
Comment (illicitonion) [inline proxy/conn_test.go:27]: t.Fatalf exists to stop the test and report an error; no need to t.Errrof ; return

File: proxy/conn_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	syncer := &testSyncer{}
+	conn := New(syncer, ws)
+
+	// call SendMessage. We don't expect it to do anything until Start() is
+	// called, though
+	msg := []byte("bobbins")
+	conn.SendMessage(msg)
+
+	// set up the expectations for the writer
+	exp := append([]byte{websocket.TextMessage | 128, byte(len(msg))}, msg...)
```
Comment (illicitonion) [inline proxy/conn_test.go:41]: 128?

File: proxy/conn_test.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
```
Comment (illicitonion) [inline proxy/conn_test.go:64]: This should probably t.Errorf rather than silently being ignored?

**Comment by richvdh** [comment]
Closing this PR for now. General feeling was that it added a lot of complexity without really testing anything useful.

**Pull request closed without merging**